latch_cli/services/cp/upload/worker.py

```diff
@@ -0,0 +1,206 @@
+                    f"{resolved}: file is {with_si_suffix(file_size)} which exceeds the"
+                    " maximum upload size (5TiB)",
+                )
+
+            chunk_size = work.chunk_size_mib * Units.MiB
+            if chunk_size < min_part_size:
+                click.secho(
+                    """\
+                    Unable to complete upload - please check your internet connection speed or any firewall settings that may block outbound traffic.\
+                    """,
+                    fg="red",
+                )
+                raise click.exceptions.Exit(1)
+
+            part_count = min(
+                latch_constants.maximum_upload_parts,
+                math.ceil(file_size / chunk_size),
+            )
+            part_size = max(
+                chunk_size,
+                math.ceil(file_size / latch_constants.maximum_upload_parts),
+            )
+
+            pbar.desc = resolved.name
+            pbar.total = file_size
+
+            # jitter to not dos nuc-data
+            await asyncio.sleep(0.1 * random.random())
+
+            resp = await sess.post(
+                "https://nucleus.latch.bio/ldata/start-upload",
+                headers={"Authorization": get_auth_header()},
+                json={
+                    "path": work.dest,
+                    "content_type": content_type,
+                    "part_count": part_count,
+                },
+            )
+            resp.raise_for_status()
+
+            json_data = await resp.json()
+            data: StartUploadData = json_data["data"]
+
+            if "version_id" in data:
+                total_pbar.update(1)
+                # file is empty - nothing to do
+                continue
+
+            try:
+                parts = await asyncio.gather(*[
+                    upload_chunk(sess, resolved, url, index, part_size, pbar)
+                    for index, url in enumerate(data["urls"])
+                ])
+            except TimeoutError:
+                await work_queue.put(
+                    Work(work.src, work.dest, work.chunk_size_mib // 2)
+                )
+
+                continue
+
+            # exception handling
+            resp = await sess.post(
+                "https://nucleus.latch.bio/ldata/end-upload",
+                headers={"Authorization": get_auth_header()},
+                json={
+                    "path": work.dest,
+                    "upload_id": data["upload_id"],
+                    "parts": [
+                        {
+                            "ETag": part.etag,
+                            "PartNumber": part.part,
+                        }
+                        for part in parts
+                    ],
+                },
+            )
+            resp.raise_for_status()
+
+            if print_file_on_completion:
+                pbar.write(work.src.name)
+
+            pbar.reset()
+            total_pbar.update(1)
+
+    pbar.clear()
+
```

Comment:
you might want to do a smarter backoff with more tries given that we can 429 on this

Reply:
this one is kinda unbounded in retries but there should be a maximum. Ideally, we have a semaphore which bounds the number of concurrent calls to nuc-data and then backoffs are less important and we can keep as is.